addon/-private/class/modifier-manager.ts

```diff
@@ -21,43 +55,76 @@ export default class ClassBasedModifierManager<S> {
       | Factory<typeof ClassBasedModifier>
       | typeof ClassBasedModifier,
     args: ArgsFor<S>
-  ): ClassBasedModifier<S> {
+  ): CreateState<S> {
     const Modifier = isFactory(factoryOrClass)
       ? factoryOrClass.class
       : factoryOrClass;
 
     const modifier = new Modifier(this.owner, args);
 
+    const state: CreateState<S> = {
+      instance: modifier,
+      element: null,
+    };
+
+    // Since we have created a state bucket to hand around, we need to make its
+    // destruction trigger the destruction of the modifier. (Ember will
+    // automatically call `destroy` on the state bucket returned from the
+    // `createModifier` hook for precisely this reason.)
+    associateDestroyableChild(state, modifier);
     registerDestructor(modifier, destroyModifier);
 
-    return modifier;
+    return state;
   }
 
   installModifier(
-    instance: ClassBasedModifier<S>,
+    state: State<S>,
```

Comment:
Yeah, the thing this “avoids” is simply creating a new object rather than reusing the old one, but a cast would also allow that, and would be more honest (if we care about avoiding the allocation).

Reply:
Resolved via ebc9285